design/ceph/object/ceph-rgw-support-multisite-for-external-cluster.md

```diff
@@ -0,0 +1,50 @@
+---
+# Improvements to RGW Multisite configuration
+target-version: release-1.10
+---
+
+# Feature Name
+ Improvements to RGW Multisite configuration
+
+## Summary
+Following are the major changes planned with this feature:
+* For external RGW servers, have place holder for zone/zonegroup/realm settings in `GatewaySpec`.
+* Provide option to add Endpoints to `zone` configuration.
+* Encode the `Endpoint` information into `realm-system-user` secret using the RGW Kubernetes service name.
+
+### Goals
+Improve usability of RGW mulitisite for customers and make it more friendly for disaster recovery scenarios. And planning to cover following use-cases:
+* two ceph object store clusters running in two different kubernetes cluster.
+* one ceph object store resides in kubernetes and another is external.
+* External ceph object store configured with custom zone/zonegroup/realms.
+
+## Proposal details
+Currently for external RGW server, Rook sets realm/zone/zonegroup in ObjContext to ""/default/default respectively. So if the RGW server belongs to custom realm/zone/zonegroup, all the `radosgw-admin` commands related may fail. So for external RGW server introducing fields to define realm/zone/zonegroup and will be part of `GatewaySpec` in `CephObjectStoreSpec`. And Rook operator will also create Kubernetes secret for storing system user details.
+```
+type ExternalMultisiteSpec struct {
```

Comment:
We have CRDs for the multisite configuration: CephObjectRealm, CephObjectZoneGroup, and CephObjectZone. Seems like we should use those instead of redefined them in the gateway.

Reply:
Didn't used CRDs since it does not generate any resources, we are just referring to existing details about RGW configuration. Rook operator is filling up Objcontext with zone/zonegroup/realm details irrespective of external or internal RGW servers. That information needs to populate correctly and it passed `radosgw-admin` commands later on.

Reply:
Same comment as Travis, I would expect each multisite CRDs to implement an external capability or something, so that `ObjContext` is filled correctly?

Reply:
@travisn @leseb how about adding `external` as `bool` in the CRDs of realm/zonegroup/zone?

Reply:
The realm has a [pull.endpoint](https://github.com/rook/rook/blob/master/deploy/examples/object-multisite-pull-realm.yaml#L23-L29) that need to be set if we are pulling from an external rgw. Does this already represent the external cluster?

Reply:
@travisn : Yaa it will work. IMO I guess the major confusion is mixing up two different features/issues together.

First issue, if RGW server part of external ceph cluster is exposed as k8s service(I mean `cephobjectstore` creation for external RGW) which has custom zone/zonegroup settings(let's leave out multisite from the picture at the moment) then all operations related such as creation/deletion/updation to `OBC`, `cephobjectstoreuser`, `cephobjectstore` needs to work properly. There are two issues related to that described in https://github.com/rook/rook/pull/10323#discussion_r887587521.

The second issue, If I need to configure up an internal add cephobjectstore to already existing external cephobjectstore. The user can manually create the secret with system user details, fetch the endpoint and give it to `realm.pull.endpoint`. What is in my mind was if the user creates(user can always choose not to) `cephobjecstore` for the rgw server in the external ceph cluster, then we can ease the manual step by encoding details into the current k8s realm secret, so that user can just refer the k8s secret for all the details

Reply:
Still trying to understand this myself. What I think I understand is that there are two cases: (1) the multisite case, and (2) the external case.

1. In the multisite case, we assume to have our Rook CephObjecStore in non-external mode. This is a CephCluster running in Kubernetes. The CephObjectRealm `spec.pull.endpoint` refers to a different Ceph object store (it may be a different Rook cluster or may be standalone Ceph, we don't care), and all the information in the realm is synchronized between the two clusters. We effectively have a local copy of an external object store.

2. The external case (that is bugged as noted in #6342) is when we have an external mode CephObjectStore. In this case we use CephObjectStore `spec.gateway.externalRgwEndpoints` to tell Rook to send all object-related commands to a given endpoint. That RGW endpoint may have any zone/zonegroup/realm configuration, and Rook needs to use that info to be able to create buckets and users on the external RGW. We are not creating a local copy of the external store, but instead we are sending all communications to the external RGW which is the only source of truth.

@thotz @alimaredia does this breakdown sound right to you two?

For the external case bugged by #6342, I have further questions:
1. Is it possible for Rook to query Ceph or the RGW via CLI/API calls to determine what zone/zonegroup/realm an RGW is in when given its `spec.gateway.externalRgwEndpoints`?
2. My experience using Ceph's CLI has been that I don't always need all that information to do the API calls. Can we get by with only zone (or only zone and realm)(or only zone and zonegroup) information?